### ISSUE TITLE
Not a Bug: Layout Editor Tool

### ISSUE THREAD

**Issue by Lixquid** [Julow/Unexpected-Keyboard#365]
Hi there!

This is Not A Bug™, just a heads-up that I've made a graphical layout editing tool at https://unexpected-keyboard-layout-editor.lixquid.com, and I'd love if you've got any feedback / any way to make the tool better for the keyboard.

I know the custom layout setting is marked as "not meant to be used", but for me allowing me to customize it exactly how I like has genuinely been what has made this the best keyboard I've ever used, and hopefully with a more user friendly tool other people will be able to experience it as well! Please, keep up the good work! 😃

(Sorry about polluting the issue tracker, but I couldn't find any other way to get in contact 🙏)

**Comment by Spike-from-NH** [comment]
That's clever!

**Comment by Julow** [comment]
This is a wonderful contribution, thanks !

There's is currently no documentation about how to use the app. But if there's ever one, this should be mentioned.

This can also be useful for people adding standard layouts to the app. I would like to add a link to your app in CONTRIBUTING.md, what do you think ?

**Comment by Spike-from-NH** [comment]
@Julow, agreed, add the link; not everyone is as comfortable hacking XML files as I am. @Lixquid, one appeal of using Github/F-Droid is open source; the user is able to see there is no hidden treachery in the app. Are you willing to expose the source code of your editor?

**Comment by Lixquid** [comment]
@Julow Thanks for taking a look! I'd be happy to add a help dialog in the top right to explain how to use the tool, and how to get layout files into / out of the app, if you think that'd help most.

@Spike-from-NH Yep, all of my web tool are open source: https://github.com/Lixquid/app-unexpected-keyboard-layout-editor

**Comment by Julow** [comment]
Added in d2a92795, thanks again !

Some remarks:
- Layouts can have more or less than 10 columns. The largest row (accounting for each key `width` and `shift`) defines the width of every keys.
- I'd expect to see a link to the source code in the app.
- The "Start from a template" dropdown seems like it would wait until I click "Import" before erasing my work but it doesn't. Perhaps it could be a modal with a "cancel" button ?
- It could autocomplete the name of keys with [special meaning](https://github.com/Julow/Unexpected-Keyboard/blob/master/srcs/juloo.keyboard2/KeyValue.java#L300).
- Corner/edge keys are represented by input boxes, each with a label, which takes a lot of space and is not intuitive. What do you think of arranging 9 input boxes in a 3x3 square, with no label ?
- Some missing options:
  + `<keyboard width="10">` to specify the width of every rows. For example, to leave empty space on the right of every rows.
  + `<row shift="0.5">` to add extra space above a row.

**Comment by Lixquid** [comment]
@Julow Thanks for the feedback! This is exactly what I was looking for when I posted this on the issue tracker 😄

I've added issues on my repository for all the points you've mentioned, I'll tackle them as I can solve them, thanks!

**Comment by Mispille** [comment]
I was thinking about making documentation for layout editing for the GitHub repo, @Lixquid would you like to help? 


@Julow would this help you?

**Comment by Julow** [comment]
Documentation would be greatly appreciated ! Thanks for contributing :)

**Comment by Mispille** [comment]
@Lixquid did a PR to change simple things, then was thinking about adding links to @Julow 's app (Play store, f-droid, GitHub ?)
Not included in the PR
